onnxruntime/core/framework/data_transfer.cc

```diff
@@ -0,0 +1,24 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/framework/data_transfer.h"
+
+namespace onnxruntime {
+
+common::Status IDataTransfer::CopyTensor(const Tensor& src, Tensor& dst) const {
+  return CopyTensor(src, dst, 0);
+}
+
+bool CPUDataTransfer::CanCopy(const OrtDevice& src_device, const OrtDevice& dst_device) const {
+  return src_device.Type() == OrtDevice::CPU && dst_device.Type() == OrtDevice::CPU;
+}
+
+common::Status CPUDataTransfer::CopyTensor(const Tensor& src, Tensor& dst, int /*exec_queue_id*/) const {
+  size_t bytes = src.DataType()->Size() * src.Shape().Size();
+  const void* src_data = src.DataRaw();
+  void* dst_data = dst.MutableDataRaw();
+  memcpy(dst_data, src_data, bytes);
```

Comment:
Do we want to skip if dst_data == src_data? Or should we assert that they are not the same nor overlapping?

Reply:
I think here would be a good place to ensure we don't do pointless copies. Maybe assert that dst.Size() <= src.Size().

---
In reply to: [297956593](https://github.com/microsoft/onnxruntime/pull/1268#discussion_r297956593) [](ancestors = 297956593)